### PR TITLE
Add CAN HIL, STM32F4xx support and capsule

### PR DESCRIPTION
**Pull request by teonaseverin**
### Pull Request Overview

This pull request adds a HIL for CAN communication, as well as a low-level driver for the STM32F4xx chips and a CAN capsule that supports single buffer message transmission and reception.

The general low-level driver is implemented in the `stm32f4xx` crate, but it is created in the specific `stm32f429zi` crate because not all stm32f4 chips support the CAN communication. The STM32F429ZI chip supports only the CAN communication protocol with messages with an 8-byte length. 

The CAN HIL exposes multiple traits, used for configuring and controlling the peripheral, as well as traits for sending receiving messages. The HIL supports both the CAN and the CANFD (CAN Flexible Data-Rate) protocol, which can communicate with flexible length messages up to 64 bytes.

### Testing Strategy

This pull request was tested on a NUCLEO-F429ZI board.

### TODO or Help Wanted

Any feedback is welcome.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: boards/components/src/can.rs
Comment (hudson-ayers) [inline boards/components/src/can.rs:45]: The updated convention is to use `static_buf!()` for all globals, even byte buffers

File: boards/components/src/can.rs
Comment (lschuermann) [inline boards/components/src/can.rs:9]: Did we reach a conclusion on when / how / whether we'd like to allow inclusion of LICENSE headers, their precise format, text and/or other conventions? I recall some discussions around these issues from some time ago, and it seems that it's important to have a consistent line of argument for our license compliance here.

**Review by phil-levis — APPROVED** [review]
Looks good to me. Just one comment: in cases when there's multiple traits with `set_client` that are sometimes combined in compound traits (like `pub trait Can` here), giving the `set_client` different names is helpful. E.g., `set_receive_client` and `set_transmit_client`. But if you prefer `set_client`s that's fine.

File: boards/components/src/can.rs
Comment (alexandruradovici) [reply boards/components/src/can.rs:9]: We do not have any requirements here expect the request of allowing us to write the "Copyright OxidOS Automotive SRL" line.

**Comment by teonaseverin** [comment]
@phil-levis I rebased the branch.

**Comment by hudson-ayers** [comment]
Sorry, this is still blocked on the outcome of the licensing discussion. Pat should be uploading a document this weekend or early next week to resolve that

**Comment by alexandruradovici** [comment]
Is there anyway we could merge this and then change the exact way of writing the copyright and license after we merge #3318, as it seems that it will take some time to reach a clear way of doing it.

**Comment by alexandruradovici** [comment]
is there any way we can merge this? It is blocked for a long time and it would really help us. Thanks.

**Comment by lschuermann** [comment]
> Is there any way we can merge this? It is blocked for a long time and it would really help us. Thanks.

I'm all for that. I believe it's fair to say that we've converged on a pretty concise notion of how we'd like to apply license headers to these files and are ironing out only the last documentation details of #3318. Worst case we can change the formatting of these changes introduced here. We've been blocking this over 2 months, which is what I wanted to avoid in the first place.

**Review by lschuermann — APPROVED** [review]
These changes look good to me! I especially appreciate the extensively documented and well-typed HIL, as well as the documented userspace driver. I don't have much experience with using CAN so I can't comment on the generalizability of this code, but I presume that as long as we don't provide any stability guarantees for this right now, we can change it when a second implementation comes along. Thanks for this contribution!

**Comment by lschuermann** [comment]
I'd assign a weak `last-call` label given this has two approvals and has been pending for so long, but hesitate to remove the `blocked` label until we have consensus that we can move forward, either by merging #3318 or otherwise.

**Review by brghena — APPROVED** [review]

File: boards/components/src/can.rs
Comment (lschuermann) [reply boards/components/src/can.rs:5]: To document the resolution of the Core WG call of 2022-01-13, we‘re able to merge this before #3318 but would appreciate if it follows the current state of examples there, i.e. has a copyright attribution to Tock contributors as well.

**Comment by alexandruradovici** [comment]
We added the copyright for Tock Contributors.

**Review by lschuermann — APPROVED** [review]
bors r+

**Review by hudson-ayers — APPROVED** [review]
bors r+

Thanks for your patience on this!

**Pull request merged**
